### PR TITLE
Preserve indentation when using `parseDocument`

### LINKED ISSUE #283: Array-indent not maintained between parseDocument/toString
**Describe the bug**

First of all, thanks for an awesome yaml parser/stringifyer! I just wrote this this weekend: [yaml-diff-patch](https://github.com/grantila/yaml-diff-patch) using this library to allow modifying a yaml using JSON patches.

I'm not sure if this is considered a bug, but I'd very much hope so. The library would be most useful if it maintains as much as possible from the source yaml when parsing the document into AST/CST so that stringifying it produces the exact same result. This means that minimal (or no) modifications to the AST would produce a minimal (or no) diff when text diffing the original yaml with the modified. In other words, **only** what you change, would be part of the final diff.

**To Reproduce**

The following:

```yaml
a:
- b: c
```

when `parseDocument` -> `toString` produces:

```yaml
a:
  - b: c
```

**Expected behaviour**

I totally understand how this looks better. I also totally understand that a document with both these indentations in different places is pretty ugly, and being able to "reformat" it is great. However, being able to maintain all blocks' indentations individually as-is makes more sense when trying to make a minimal diff that only touches what it needs.

I noticed (in 1.10.2) that quite a few things get re-organized, long lines get chopped into multiple shorter, long one-line arrays get split into line-by-line arrays etc. I haven't tried that in 2.0 yet, so you can ignore that for now.

**Versions (please complete the following information):**

- Environment: Node.js 14.15.1
- `yaml`: 1.10.2 and 2.0.0-6

### PR DESCRIPTION
**Pull request by Nathan-Fenner**
This fixes #283 (with some qualifications).

Documents like the following now get their indentation preserved:

```yaml
foo0:
- a
- b
- c
foo2:
  - a
  - b
  - c
foo8:
        - a
        - b
        - c
bar:
       more:
                     indentation:
                                             is: needed
```

when you enable the new option `preserveCollectionIndentation` and also enable the `keepSourceTokens` option:

```ts
const document = parseDocument(yamlContents, {
  keepSourceTokens: true,
  preserveCollectionIndentation: true,
});
```

If you insert/edit/move around nodes in this document, or insert them into others, their indentation is also preserved.

---

I added the new option `preserveCollectionIndentation` purely as a backwards compatibility measure. It could also default to `true`, in which case you'd only need to enable `keepSourceTokens` for it to work.

Unfortunately at least with the way the parser is structured right now, we do need the source tokens since they contain the relevant indentation information that makes the preservation possible.

---

I added a new test similar to the example above which ensures that the document round-trips exactly, maintaining all indentation.

I also added a test that verifies all documents in the yaml-suite can be round-tripped twice to verify that this doesn't break them (not that we'd expect that).

The one special edge case is that if we have a document like

```yaml
list:
- not indented
- at all
```

and we replace the `list` with an object like `{greeting: "hello"}` as in:
```ts
const doc = parseDocument(yamlContents);
doc.set('list', {greeting: "hello"});

console.info(doc.toString());
```
then we need to insert some new indentation; lists are allowed to have an `indentStep` of `0` since the hyphen marks that the item starts, but non-lists are not:

```
# right
list:
  greeting: hello

# wrong (missing indentation changes meaning of document)
list:
greeting: hello
```

in this case, we default back to the configured `indentStep` (e.g. 2).

### CODE REVIEW

**Review by eemeli — COMMENTED** [review]
This looks really promising! I left a couple of line comments below, but I'll need to spend a bit more time on a proper review at some point.

My main initial concerns are:
1. Is this a good idea to include as a partial implementation first, or do we need to ensure that it also covers block sequences and flow collections?
2. What happens when a new key/value pair is added to a map that has pairs with a set `srcIndentStep`? Could/should the value (also?) be set on the collection rather than the pair?

File: src/compose/resolve-block-map.ts [hunk 1]
```diff
@@ -102,7 +102,24 @@ export function resolveBlockMap(
...
+        const keyIndent: number | undefined = (collItem?.key as any)?.indent
+        const valueIndent: number | undefined = (collItem?.value as any)?.indent
+
+        if (
+          ctx.options.preserveCollectionIndentation &&
+          keyIndent !== undefined &&
+          valueIndent !== undefined &&
+          valueIndent >= keyIndent
+        ) {
+          pair.srcIndentStep = valueIndent - keyIndent
```
[the remark this answers is not in the document]
Comment (Nathan-Fenner) [reply src/compose/resolve-block-map.ts:120]: Huh, for some reason in my initial tests I thought that we wouldn't have the right source indent info without `keepSourceTokens` enabled. So, this is totally unnecessary!

File: src/compose/resolve-block-map.ts [hunk 2]
```diff
@@ -102,7 +102,24 @@ export function resolveBlockMap(
...
+        // If so, we can compute the difference between them and preserve it
+        // if the preserveCollectionIndentation option is set.
+        const keyIndent: number | undefined = (collItem?.key as any)?.indent
+        const valueIndent: number | undefined = (collItem?.value as any)?.indent
+
+        if (
+          ctx.options.preserveCollectionIndentation &&
+          keyIndent !== undefined &&
+          valueIndent !== undefined &&
+          valueIndent >= keyIndent
```
[the remark this answers is not in the document]
Comment (Nathan-Fenner) [reply src/compose/resolve-block-map.ts]: TypeScript doesn't permit you to compare `number|undefined` values with `>=`, even though JavaScript has the right semantics. So we either need a cast (which seems unnecessary) or we need to explicitly include the logic to exclude `undefined` values.

[the remark this answers is not in the document]
Comment (Nathan-Fenner) [reply src/compose/resolve-block-map.ts]: I've fixed this to avoid the `as any` casts; I didn't notice the destructuring in the first past. Now it's simplified (somewhat) to

```ts
const keyIndent: number | undefined =
  key && 'indent' in key ? key.indent : undefined
const valueIndent: number | undefined =
  value && 'indent' in value ? value.indent : undefined
```

The extra logic here is needed to convince TypeScript that it's okay:

- `key.indent` is only valid if we know that `"indent"` is in `key` in the first place (and some of the variants don't include an `indent` field), so we have to do `"indent" in key` before we can write `key.indent`
- `'indent' in key` is only valid if `key` is not null/undefined, so we have to check `key &&` first

**Comment by Nathan-Fenner** [comment]
> Is this a good idea to include as a partial implementation first, or do we need to ensure that it also covers block sequences and flow collections?

It handles block sequences in the sense of knowing how much indentation occurs _before_ a `-` which is the common case when used as a list in a map (or a list of fields), see the updated example in the PR description.

The less common case is where you have indentation after the `-`, as in:

```
-
  -
    - foo
    - bar
  -
    - baz
```

Similarly, it reformats

```yaml
list_map2:
  -
    foo: 1
    bar: 2
    qux: 3
  -
    foo: 4
    bar: 5
    qux: 6
```

into

```yaml
list_map2:
  - foo: 1
    bar: 2
    qux: 3
  - foo: 4
    bar: 5
    qux: 6
```

Since these kinds of structures are less common (in my experience) I think it's fine as a first step to exclude them, and then revisit the approach in a later PR.

---

Similarly, I'm skipping over flow collections, since they're just fairly different. They have different kinds of indentation/spacing concerns so the implementation would probably be fairly different. And in my experience they're usually much smaller e.g. single-line, with little variation in how they're formatted, so they're more likely to just be preserved anyway.

---

> What happens when a new key/value pair is added to a map that has pairs with a set srcIndentStep? Could/should the value (also?) be set on the collection rather than the pair?

My main goal in this change is to avoid making unnecessary edits to the original, rather than trying to "preserve style". So in this case, if a new "block" is made, that block will have a regular indent step (e.g. 2). So for example, if we had the following document:

```yaml
count:
          animals:
                    cat: 1
                    dog: 2
```

if we write `doc.setIn(["count", "animals", "lizard"], 3)` then we get

```yaml
count:
          animals:
                    cat: 1
                    dog: 2
                    lizard: 3
```
because the indentation before `lizard` belongs to the `Pair` with `key` `animals`. So this behaves exactly as we expect.

On the other hand, if we have

```yaml
count:
        mammals:
                cat: 1
                dog: 2
```

and we write `doc.setIn(["count", "reptiles"], {"lizard": 3, "snake": 4});` then we get

```yaml
count:
        mammals:
                cat: 1
                dog: 2
        reptiles:
          lizard: 3
          snake: 4
```

So the yaml is still valid, but the new block is not indented as much as its sibling blocks. I think that there is a case to be made that the parent block could infer a "desirable" step for new children pairs, but as far as addressing my original goal of eliminating redundant diffs, it makes no difference. So I'm ambivalent about including this now, since it might be nice, but it would be more work and doesn't directly help my use-case.

File: src/compose/resolve-block-map.ts [hunk 3]
```diff
@@ -102,7 +102,27 @@ export function resolveBlockMap(
...
+        pair.srcToken = collItem
+      }
+
+      // Check to see if the key tokens exist and have an indentation set.
+      // If so, we can compute the difference between them and preserve it
+      // if the preserveCollectionIndentation option is set.
+      const keyIndent: number | undefined =
+        key && 'indent' in key ? key.indent : undefined
+      const valueIndent: number | undefined =
+        value && 'indent' in value ? value.indent : undefined
```
Comment (eemeli) [inline src/compose/resolve-block-map.ts:115]: It would probably be clearer to skip defining these intermediate variables, and just use the conditionals in the subsequent if statement directly.

**Review by eemeli — CHANGES_REQUESTED** [review]
Sorry to have kept you waiting; been focusing on other projects recently.

On further thought, I think a solution is needed here for keeping all map values similarly indented when they change. As in, if this document is parsed with `keepMapIndent: true`:

```yaml
foo: |
    bar
x: y
```
and then the `y` value is updated to `'y\nz\n'`, then the stringified output should be:
```yaml
foo: |
    bar
x: |
    y
    z
```

That probably means that something like a `srcIndentSep` value also needs to be set on a block map during its composition.

The added test cases should also cover explicit tags and aliases, complex/explicit keys, as well as block scalars and flow collection values. All of those present occasionally interesting challenges for YAML indentation.

**Comment by eemeli** [comment]
Closing due to inactivity.
